File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/PokemonsViewModel.java
package com.mdgd.pokemon.ui.pokemons;

import androidx.core.util.Pair;
import androidx.lifecycle.Lifecycle;

import com.mdgd.mvi.MviViewModel;
import com.mdgd.pokemon.models.filters.CharacteristicComparator;
import com.mdgd.pokemon.models.filters.FilterData;
import com.mdgd.pokemon.models.filters.StatsFilter;
import com.mdgd.pokemon.models.infra.Result;
import com.mdgd.pokemon.models.repo.PokemonsRepo;
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema;
import com.mdgd.pokemon.ui.pokemons.adapter.Pokemon;
import com.mdgd.pokemon.ui.pokemons.state.PokemonsScreenEffect;
import com.mdgd.pokemon.ui.pokemons.state.PokemonsScreenState;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.schedulers.Schedulers;
import io.reactivex.rxjava3.subjects.BehaviorSubject;
import io.reactivex.rxjava3.subjects.PublishSubject;

public class PokemonsViewModel extends MviViewModel<PokemonsContract.View, PokemonsScreenState> implements PokemonsContract.ViewModel {

    private final PublishSubject<Integer> loadPageSubject = PublishSubject.create();
    private final BehaviorSubject<FilterData> filtersSubject = BehaviorSubject.createDefault(new FilterData());
    private final StatsFilter statsFilters;
    private final PokemonsRepo repo;

    public PokemonsViewModel(PokemonsRepo repo, StatsFilter statsFilters) {
        this.repo = repo;
        this.statsFilters = statsFilters;
    }

    @Override
    public void onStateChanged(Lifecycle.Event event) {
        super.onStateChanged(event);
        if (event == Lifecycle.Event.ON_CREATE && !hasOnDestroyDisposables()) {
            observeTillDestroy(
                    loadPageSubject
                            .doOnNext(ignore -> setState(new PokemonsScreenState.SetProgressVisibility(true)))
                            .switchMapSingle(page -> repo.getPage(page)
                                    .subscribeOn(Schedulers.io())
                                    .observeOn(AndroidSchedulers.mainThread())
                                    .map(result -> new Pair<>(page, result))
                            )
                            .subscribe(this::handleResult, e -> setEffect(new PokemonsScreenEffect.ShowErrorEffect(e))),

                    // do we need to sort list once more when there is a filter and new page arrived?
                    filtersSubject
                            .map(filters -> sort(filters, mapToDto(repo.getPokemons())))
                            .subscribe(this::setState, e -> setEffect(new PokemonsScreenEffect.ShowErrorEffect(e)))

            );
            loadPageSubject.onNext(0);
        }
    }

    private List<Pokemon> mapToDto(List<PokemonFullDataSchema> pokemons) {
        final List<Pokemon> items = new ArrayList<>(pokemons.size());
        for (PokemonFullDataSchema i : pokemons) {
            items.add(new Pokemon(i));
        }
        return items;
    }

    private PokemonsScreenState sort(FilterData filters, List<Pokemon> pokemons) {
        // potentially, we can create a custom list of filters in separate model. In UI we can show them in recyclerView
        if (!filters.isEmpty()) {
            final Map<String, CharacteristicComparator> comparatorMap = statsFilters.getFilters();
            Collections.sort(pokemons, (pokemon1, pokemon2) -> {
                int compare = 0;
                for (String filter : filters.getFilters()) {
                    if (comparatorMap.get(filter) != null) {
                        compare = comparatorMap.get(filter).compare(pokemon2.schema, pokemon1.schema); // swap, instead of multiply on -1
                        if (compare != 0) {
                            break;
                        }
                    }
                }
                return compare;
            });
        }
        return new PokemonsScreenState.SetItems(pokemons, statsFilters.getAvailableFilters());
    }

    private void handleResult(Pair<Integer, Result<List<PokemonFullDataSchema>>> pair) {
        if (pair.second.isError()) {
            setEffect(new PokemonsScreenEffect.ShowErrorEffect(pair.second.getError()));
        } else {
            final List<Pokemon> list = mapToDto(pair.second.getValue());
            if (pair.first == 0) {
                setState(new PokemonsScreenState.SetItems(list, statsFilters.getAvailableFilters()));
            } else {
                setState(new PokemonsScreenState.AddItems(list));
            }
        }
    }

    @Override
    public void reload() {
        loadPageSubject.onNext(0);
    }

    @Override
    public void loadPage(int page) {
        loadPageSubject.onNext(page);
    }

    @Override
    public void sort(String filter) {
        final List<String> filters = new ArrayList<>(3);
        if (getState() != null && getState().getActiveFilters() != null) {
            filters.addAll(getState().getActiveFilters());
        }
        if (filters.contains(filter)) {
            filters.remove(filter);
        } else {
            filters.add(filter);
        }
        setState(new PokemonsScreenState.ChangeFilterState(filters));
        filtersSubject.onNext(new FilterData(filters));
    }

    @Override
    public void onItemClicked(PokemonFullDataSchema pokemon) {
        setEffect(new PokemonsScreenEffect.ShowDetails(pokemon.getPokemonSchema().getId()));
    }
}

File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/PokemonsViewModel.kt
package com.mdgd.pokemon.ui.pokemons

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.viewModelScope
import com.mdgd.mvi.MviViewModel
import com.mdgd.pokemon.models.filters.FilterData
import com.mdgd.pokemon.models.filters.StatsFilter
import com.mdgd.pokemon.models.repo.PokemonsRepo
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema
import com.mdgd.pokemon.models.util.DispatchersHolder
import com.mdgd.pokemon.ui.pokemons.state.PokemonsScreenEffect
import com.mdgd.pokemon.ui.pokemons.state.PokemonsScreenState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class PokemonsViewModel @Inject constructor(
    private val repo: PokemonsRepo,
    private val filtersFactory: StatsFilter,
    private val dispatchers: DispatchersHolder
) : MviViewModel<PokemonsContract.View, PokemonsScreenState, PokemonsScreenEffect>(),
    PokemonsContract.ViewModel {

    private var firstVisibleIndex: Int = 0
    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
        setEffect(PokemonsScreenEffect.Error(e))
    }
    private val pageFlow = MutableStateFlow(0)
    private val filterFlow = MutableStateFlow(FilterData())
    private var launch: Job? = null

    override fun onStateChanged(event: Lifecycle.Event) {
        super.onStateChanged(event)
        if (event == Lifecycle.Event.ON_CREATE && launch == null) {
            launch = viewModelScope.launch(exceptionHandler) {
                pageFlow
                    .onEach { setState(PokemonsScreenState.Loading()) }
                    .flowOn(dispatchers.getMain())
                    .map { page -> Pair(page, repo.getPage(page)) }
                    .flowOn(dispatchers.getIO())
                    .catch { e: Throwable -> setEffect(PokemonsScreenEffect.Error(e)) }
                    .collect { pagePair: Pair<Int, List<PokemonFullDataSchema>> ->
                        if (pagePair.first == 0) {
                            setState(
                                PokemonsScreenState.SetData(
                                    pagePair.second, filtersFactory.getAvailableFilters()
                                )
                            )
                        } else {
                            setState(PokemonsScreenState.AddData(pagePair.second))
                        }
                    }
            }

            viewModelScope.launch(exceptionHandler) {
                filterFlow
                    .map { sort(it, repo.getPokemons()) }
                    .collect { sortedList ->
                        firstVisibleIndex = 0
                        setState(PokemonsScreenState.UpdateData(sortedList))
                        setEffect(PokemonsScreenEffect.ScrollToStart())
                    }
            }
        }
    }

    private fun sort(
        filters: FilterData, pokemons: List<PokemonFullDataSchema>
    ): List<PokemonFullDataSchema> {
        val list = ArrayList(pokemons)
        if (!filters.isEmpty) {
            val comparators = filtersFactory.getFilters()
            list.sortWith { pokemon1: PokemonFullDataSchema?, pokemon2: PokemonFullDataSchema? ->
                var compare = 0
                for (filter in filters.filters) {
                    // swap, instead of multiply on -1
                    compare = comparators[filter]?.compare(pokemon2!!, pokemon1!!) ?: 0
                    if (compare != 0) {
                        break
                    }
                }
                compare
            }
        }
        return list
    }

    override fun reload() {
        pageFlow.tryEmit(0)
    }

    override fun sort(filter: String) {
        val filters = getState()?.activeFilters?.toMutableList() ?: mutableListOf()
        if (filters.contains(filter)) {
            filters.remove(filter)
        } else {
            filters.add(filter)
        }
        setState(PokemonsScreenState.ChangeFilterState(filters))
        filterFlow.tryEmit(FilterData(filters))
    }

    override fun onItemClicked(pokemon: PokemonFullDataSchema) {
        setEffect(PokemonsScreenEffect.ShowDetails(pokemon.pokemonSchema?.id))
    }

    override fun onScroll(firstVisibleIndex: Int, lastVisibleIndex: Int) {
        this.firstVisibleIndex = firstVisibleIndex
        val page = pageFlow.value + 1
        if (lastVisibleIndex >= page * PokemonsRepo.PAGE_SIZE - 8) {
            pageFlow.tryEmit(page)
        }
    }

    override fun firstVisible() = firstVisibleIndex

    public override fun onCleared() {
        launch = null
        super.onCleared()
    }
}
